File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        result = []
        
        hashMap = dict()
        max_num = float('-inf')
        
        for num in nums:
            if num not in hashMap.keys():
                hashMap[num] = 1
            else:
                hashMap[num] = hashMap.get(num) + 1
            
            max_num = max(max_num, hashMap.get(num))
        
        bucketArr = [None for i in range(max_num + 1)]
        
        for key, value in hashMap.items():
            if bucketArr[value]:
                bucketArr[value].append(key)
            else:
                bucketArr[value] = [key]
        
        for i in range(len(bucketArr) - 1, 0, -1):
            if k > 0:
                curr = bucketArr[i]                
                if bucketArr[i]:
                    for c in range(len(curr)):
                        if k > 0:
                            result.append(curr[c])
                            k -= 1
                        else:
                            break
            else:
                break
        # print(bucketArr, k)
        return result

#HashMap + BuckerArr
#Time Complexity: O(2n)
#Space Complexity: O(n)

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
import java.util.List;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        if(nums == null || nums.length == 0){
            return new int[]{};
        }
        
        int[] res = new int[k];
        int max = Integer.MIN_VALUE;
        HashMap<Integer, Integer> map = new HashMap<>();
        
        for(int num: nums){
            map.put(num, map.getOrDefault(num, 0) + 1);
            max = Math.max(max, map.get(num));            
        }
        
        List [] bucketArr = new List[max + 1];
        

        for(int key: map.keySet()){
            int index = map.get(key);
            if(bucketArr[index] == null){
                bucketArr[index] = new ArrayList<>();
            }
                
            bucketArr[index].add(key);
            
    }   
    
        for(int i = max; i >= 0 && k > 0; i--){
            if(bucketArr[i] != null){
                List<Integer> list = bucketArr[i];
                for(int j=0; j< bucketArr[i].size() && k>0; j++){
                    res[k- 1] = list.get(j);
                    k--;
                }
            }
        }
        
        return res;
    }
}